driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -203,64 +245,97 @@ public void call(final AsyncConnection connection, final Throwable t1) {
+    private BulkWriteResult executeBulkWriteBatch(final WriteBinding binding, final Connection connection,
+                                                  final BulkWriteBatch originalBatch) {
+        BulkWriteBatch currentBatch = originalBatch;
+        MongoException exception = null;
 
-        while (batch.shouldProcessBatch()) {
-            BsonDocument result = connection.command(namespace.getDatabaseName(), batch.getCommand(), NO_OP_FIELD_NAME_VALIDATOR,
-                    ReadPreference.primary(), batch.getDecoder(), sessionContext, shouldAcknowledge(batch, writeConcern),
-                    batch.getPayload(), batch.getFieldNameValidator());
-            batch.addResult(result);
-            batch = batch.getNextBatch();
+        try {
+            while (currentBatch.shouldProcessBatch()) {
+                BsonDocument result = executeCommand(connection, currentBatch, binding);
+                currentBatch.addResult(result);
+                currentBatch = currentBatch.getNextBatch();
+            }
+        } catch (MongoException e) {
+            exception = e;
+        } finally {
+            connection.release();
         }
 
-        return batch.getResult();
+        if (exception == null) {
+            return currentBatch.getResult();
+        } else if (!currentBatch.shouldProcessBatch() || !originalBatch.getRetryWrites() || !isRetryableException(exception)) {
+            throw exception;
+        } else {
+            return retryExecuteBatches(binding, currentBatch, exception);
+        }
     }
 
-    private BulkWriteResult executeLegacyBatches(final Connection connection) {
-        for (WriteRequest writeRequest : getWriteRequests()) {
-            if (writeRequest.getType() == INSERT) {
-                connection.insert(getNamespace(), isOrdered(), (InsertRequest) writeRequest);
-            } else if (writeRequest.getType() == UPDATE || writeRequest.getType() == REPLACE) {
-                connection.update(getNamespace(), isOrdered(), (UpdateRequest) writeRequest);
-            } else {
-                connection.delete(getNamespace(), isOrdered(), (DeleteRequest) writeRequest);
+    private BulkWriteResult retryExecuteBatches(final WriteBinding binding, final BulkWriteBatch retryBatch,
+                                                final MongoException originalError) {
+        return withReleasableConnection(binding, new CallableWithConnectionAndSource<BulkWriteResult>() {
+            @Override
+            public BulkWriteResult call(final ConnectionSource source, final Connection connection) {
+                if (!isRetryableWrite(retryWrites, writeConcern, source.getServerDescription(), connection.getDescription())) {
+                    connection.release();
+                    throw originalError;
+                } else {
+                    try {
+                        retryBatch.addResult(executeCommand(connection, retryBatch, binding));
```

Comment:
Stepping through this, I noticed that the payload has its position set to non-zero on the retry.  It works out fine since CommandMessage always starts at 0 and ignores the current position, but I wasn't expecting it.  No change requested, just an FYI in case you weren't aware.

Reply:
Thats expected as its the `BsonWriter` that sets the payload position as it writes it. I didn't see the need to reset it as you mentioned the current position is only needed if the payload actually needs to be split.